packages/search-in-workspace/src/browser/quick-search-in-workspace.ts

```diff
@@ -0,0 +1,230 @@
+                this.cancelCurrentSeach();
+
+                if (lookFor.length === 0) {
+                        /* The user has emptied the search box, call acceptor to
+                           remove any previously shown results.  */
+                        acceptor([]);
+                        return;
+                }
+
+                /* Options passed to the search service.  */
+                const opts: SearchInWorkspaceOptions = {
+                        maxResults: this.MAX_RESULTS,
+                };
+
+                /* The array in which we'll keep accumulating search results.  */
+                const items: QuickSearchInWorkspaceResultItem[] = [];
+
+                this.searchInWorkspaceService.search(lookFor, {
+
+                        onResult: (searchId: number, result: SearchInWorkspaceResult) => {
+                                /* Is this result from a previous search?  */
+                                if (searchId !== this.currentSearchId) {
+                                        return;
+                                }
+
+                                items.push(new QuickSearchInWorkspaceResultItem(result, this.openerService, this.labelProvider));
+                        },
+
+                        onDone: (searchId: number, error?: string) => {
+                                if (searchId !== this.currentSearchId) {
+                                        this.logger.debug('Search ' + this.currentSearchId + ' has completed, but it\'s not the current search.');
+                                        return;
+                                }
+                                this.logger.debug('Search ' + this.currentSearchId + ' has completed and is the current search.');
+                                this.currentSearchId = -1;
+
+                                if (error) {
+                                        this.showFakeResult(error, acceptor);
+                                } else if (items.length !== 0) {
+                                        items.sort((a, b) => SearchInWorkspaceResultCompare(a.getResult(), b.getResult()));
+                                        acceptor(items);
+                                } else {
+                                        this.showFakeResult('No matches :(', acceptor);
+                                }
+
+                        },
+                }, opts).then(searchId => {
+                        this.currentSearchId = searchId;
+                });
+        }
+
+        showFakeResult(label: string, acceptor: (items: QuickOpenItem[]) => void) {
+                acceptor([
+                        new QuickOpenItem({
+                                label: label,
+                        }),
+                ]);
+        }
+
+        /* If we have an ongoing search, cancel it.  */
+        cancelCurrentSeach() {
+                if (this.currentSearchId >= 0) {
```

Comment:
Instead of exposing and managing the process numbers, we should use the cancellationToken.

Reply:
Actually, I think using the ids is fine. I understand now that with all the async nature and potentially multiple clients, we need to make sure that we can associate matches with an actual search. So please ignore the comment above.